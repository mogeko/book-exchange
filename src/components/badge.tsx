import { MessageContext } from "@/providers/msgProvider";
import { useContext } from "react";

const Badge: React.FC<BadgeProps> = ({ badgeKey, className }) => {
  const msg = useContext(MessageContext);
  const count = msg.filter((m) => m.key === badgeKey).length;
  return count ? <span className={`badge ${className}`}>{count}</span> : null;
};

interface BadgeProps {
  badgeKey: string;
  className?: string;
}

export default Badge;
