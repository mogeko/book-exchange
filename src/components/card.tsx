import { type BookTypes } from "@/lib/hooks/useBooks";
import Skeleton from "@/components/skeleton";
import StarsRate from "@/components/stars";
import Image from "next/image";
import Link from "next/link";

const CardRoot: React.FC<CardProps> = ({ title, cover, author, id }) => {
  return (
    <div className="relative flex flex-col w-28">
      <figure className="flex justify-center items-center">
        <Link href={`/books/${id}`}>
          <a className="relative w-28 h-[calc(7rem/2*3)]">
            <Image src={cover!} layout="fill" alt={title} />
          </a>
        </Link>
      </figure>
      <div className="flex flex-col text-sm">
        <Link href={`/books/${id}`}>
          <a className="link no-underline text-primary hover:bg-primary hover:text-primary-content focus:bg-primary-focus focus:text-primary-content">
            <h2 className="truncate">{title}</h2>
          </a>
        </Link>
        <p className="truncate">{author}</p>
      </div>
    </div>
  );
};

const CardSkeleton: React.FC = () => (
  <Skeleton.Pulse className="relative flex flex-col mb-4 w-28">
    <Skeleton.Square className="mb-2 w-28 h-[calc(7rem/2*3)]" />
    <Skeleton.Line className="mb-2 w-24" />
    <Skeleton.Line className="mb-2 w-20" />
  </Skeleton.Pulse>
);

const LongCardRoot: React.FC<CardProps> = (props) => {
  const { title, cover, author, id, description, rates } = props;
  return (
    <div className="flex py-4 gap-3">
      <figure className="flex justify-center items-center">
        <Link href={`/books/${id}`}>
          <a className="relative w-28 h-[calc(7rem/2*3)]">
            <Image src={cover!} layout="fill" alt={title} />
          </a>
        </Link>
      </figure>
      <div className="flex flex-col gap-2 h-[calc(7rem/2*3)]">
        <div className="inline-flex justify-start">
          <Link href={`/books/${id}`}>
            <a className="link no-underline text-primary hover:bg-primary hover:text-primary-content focus:bg-primary-focus focus:text-primary-content">
              <h2 className="truncate">{title}</h2>
            </a>
          </Link>
        </div>
        <p className="truncate">{author}</p>
        <StarsRate rates={rates!} />
        <p className="line-clamp-3 text-ellipsis">{description}</p>
      </div>
    </div>
  );
};

const LongCardSkeleton: React.FC = () => (
  <Skeleton.Pulse className="relative flex mb-4 w-full gap-3">
    <Skeleton.Square className="mb-2 w-28 h-[calc(7rem/2*3)]" />
    <div className="flex flex-col w-full gap-2">
      <Skeleton.Line className="mb-2 w-1/5" />
      <Skeleton.Line className="mb-2 w-1/6" />
      <div>
        <Skeleton.Line className="mb-2 w-4/5" />
        <Skeleton.Line className="mb-2 w-4/5" />
        <Skeleton.Line className="mb-2 w-4/5" />
        <Skeleton.Line className="mb-2 w-4/6" />
      </div>
    </div>
  </Skeleton.Pulse>
);

const Card = Object.assign(CardRoot, { Skeleton: CardSkeleton });
export const LongCard = Object.assign(LongCardRoot, {
  Skeleton: LongCardSkeleton,
});

type CardProps = BookTypes;

export default Card;
