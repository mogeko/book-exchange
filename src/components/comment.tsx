import Editor from "@/components/editor/editor";

const Comments: React.FC = () => {
  return <Editor />;
};

export default Comments;
