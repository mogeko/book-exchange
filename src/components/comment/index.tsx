import CommentCard from "@/components/comment/commentCard";
import useComments from "@/lib/hooks/useComments";
import Editor from "@/components/editor";
import fetcher from "@/lib/fetcher";
import type { EditorFormInput } from "@/components/editor";
import type { CommentType } from "@/lib/hooks/useComments";

const Comments: React.FC<CommentsProps> = ({ id }) => {
  const { data, isLoading } = useComments({ bkid: id });
  const handleSubmit = async (data: EditorFormInput) => {
    await fetcher("/api/comments/submit", {
      method: "POST",
      body: JSON.stringify(data),
      headers: {
        "Content-Type": "application/json",
      },
    });
  };

  return (
    <div className="flex flex-col gap-4 max-w-xl">
      <Editor onSubmit={handleSubmit} />
      {isLoading ? (
        <CommentCard.Skeleton />
      ) : (
        data?.map((comment: CommentType, index) => (
          <CommentCard key={index} comment={comment} />
        ))
      )}
    </div>
  );
};

interface CommentsProps {
  id: `bk${number}`;
}

export default Comments;
