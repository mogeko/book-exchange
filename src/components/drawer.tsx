import { HiMenu } from "react-icons/hi";
import Link from "next/link";
import useMenus from "@/lib/connect/menus";

const DrawerMenu: React.FC<DrawerMenuProps> = ({ toggleId }) => {
  const { menus, isLoading, isError } = useMenus("/api/menus/root.json");
  if (isLoading) return <div>Loading...</div>;
  if (isError) return <div>Error...</div>;
  return (
    <div className="drawer-side">
      <label htmlFor={toggleId} className="drawer-overlay" />
      <ul className="menu p-4 overflow-y-auto w-80 bg-base-200 text-base-content">
        {menus?.data.map((menu, index) => (
          <li key={index}>
            <Link href={menu.href}>
              <a className="justify-between">
                {menu.name}
                {menu.badge ? (
                  <span className="badge">{menu.badge}</span>
                ) : null}
              </a>
            </Link>
          </li>
        ))}
      </ul>
    </div>
  );
};

export const withDrawer = (
  WrappedComponent: React.FC<Pick<DrawerProps, "children">>
) => {
  const DrawerWrapper: React.FC<DrawerProps> = (props) => {
    return (
      <div className="drawer">
        <input id={props.toggleId} type="checkbox" className="drawer-toggle" />
        <div className="drawer-content">
          <WrappedComponent>{props.children}</WrappedComponent>
        </div>
        <DrawerMenu toggleId={props.toggleId} />
      </div>
    );
  };
  return DrawerWrapper;
};

export const DrawerButton: React.FC<DrawerButtonProps> = ({ toggleId }) => {
  return (
    <label htmlFor={toggleId} className="btn btn-ghost btn-circle">
      <HiMenu className="h-6 w-6" aria-hidden="true" />
    </label>
  );
};

interface DrawerMenuProps {
  toggleId: string;
}

type DrawerProps = {
  children: React.ReactNode;
} & DrawerMenuProps;

type DrawerButtonProps = Pick<DrawerProps, "toggleId">;
