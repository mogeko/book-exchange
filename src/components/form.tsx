import { useForm } from "react-hook-form";
import type { UseFormRegister, UseFormProps } from "react-hook-form";
import type { FieldValues, SubmitHandler, FieldErrors } from "react-hook-form";
import React from "react";
import classNames from "@/lib/utils/classNames";

const Form: React.FC<FormProps> = (props) => {
  const { children, onSubmit, className, loading, opts } = props;
  const { handleSubmit, ...methods } = useForm(opts);
  const { errors } = methods.formState;

  return (
    <form className={className} onSubmit={handleSubmit(onSubmit)}>
      {React.Children.map(children, (child) => {
        if (!child.props.name && !child.props.label) return child;
        return React.createElement(child.type, {
          ...child.props,
          ...(loading ? { disabled: true } : {}),
          key: child.props.name ?? child.props.label,
          register: methods.register,
          errors: errors,
        });
      })}
    </form>
  );
};

export const Input: React.FC<InputProps> = (props) => {
  const { label, name, register, className, size = "md", ...rest } = props;
  const error = rest.errors?.[name ?? label?.toLowerCase()!]?.message;

  if (register && (name || label)) {
    return (
      <div className="form-control">
        {label && (
          <label className="label">
            <span className="label-text">{label}</span>
          </label>
        )}
        <input
          className={
            className ??
            classNames(
              "input input-bordered",
              error ? "input-error" : "",
              `input-${size}`
            )
          }
          placeholder={rest.placeholder ?? label?.toLowerCase() ?? name}
          {...register(name ?? label?.toLowerCase()!)}
          {...rest}
        />
        {error && (
          <label className="label">
            <span className="label-text-alt text-error">{error as any}</span>
          </label>
        )}
      </div>
    );
  }
  return <input className={className ?? "input input-bordered"} {...rest} />;
};

interface FormProps {
  children: React.ReactElement<FormChild>[];
  onSubmit: SubmitHandler<any>;
  opts?: UseFormProps<FieldValues>;
  className?: string;
  loading?: boolean;
}

type FormChild = {
  register?: UseFormRegister<FieldValues>;
  errors?: FieldErrors<FieldValues>;
  size?: "xs" | "sm" | "md" | "lg";
  label?: string;
  name?: string;
  className?: string;
};

type InputProps = FormChild & React.InputHTMLAttributes<HTMLInputElement>;

export default Form;
