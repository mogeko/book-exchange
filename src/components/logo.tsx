import Image from "next/image";
import Link from "next/link";

const Logo: React.FC<LogoProps> = ({ href, src, children }) => {
  return (
    <div className="flex justify-start">
      <Link href={href ?? "/"}>
        <a className="btn btn-ghost normal-case">
          <span className="sr-only">{children ?? "Logo"}</span>
          {src ? (
            <Image className="h-8 w-auto sm:h-10" src={src} alt="logo" />
          ) : (
            <h1 className="text-xl font-bold text-primary-content">
              {children}
            </h1>
          )}
        </a>
      </Link>
    </div>
  );
};

type LogoProps = {
  href?: string;
  children?: React.ReactNode;
} & Partial<Pick<Parameters<typeof Image>[0], "src">>;

export default Logo;
