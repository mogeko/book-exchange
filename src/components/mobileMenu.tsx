import { HiMenu } from "react-icons/hi";
import Link from "next/link";

const Drawer: React.FC<DrawerProps> = ({ data, children, toggleId }) => {
  return (
    <div className="drawer">
      <input id={toggleId} type="checkbox" className="drawer-toggle" />
      <div className="drawer-content">{children}</div>
      <div className="drawer-side">
        <label htmlFor={toggleId} className="drawer-overlay" />
        <ul className="menu p-4 overflow-y-auto w-80 bg-base-200 text-base-content">
          {data.map((menu, index) => (
            <li key={index}>
              <Link href={menu.href}>
                <a className="justify-between">
                  {menu.name}
                  {menu.badge ? (
                    <span className="badge">{menu.badge}</span>
                  ) : null}
                </a>
              </Link>
            </li>
          ))}
        </ul>
      </div>
    </div>
  );
};

const Button: React.FC<ButtonProps> = ({ toggleId }) => {
  return (
    <label htmlFor={toggleId} className="btn btn-ghost btn-circle">
      <HiMenu className="h-6 w-6" aria-hidden="true" />
    </label>
  );
};

const MobileMenuDrawer = Object.assign(Drawer, { Button });

interface DrawerProps {
  data: {
    name: string;
    href: string;
    badge?: number;
  }[];
  children: React.ReactNode;
  toggleId: string;
}

type ButtonProps = Pick<DrawerProps, "toggleId">;

export default MobileMenuDrawer;
