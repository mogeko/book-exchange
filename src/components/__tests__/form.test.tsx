import "@testing-library/jest-dom";
import { render, screen, renderHook } from "@testing-library/react";
import { Input } from "@/components/form";
import { useForm } from "react-hook-form";

describe("Input", () => {
  it("renders an <input> with label", () => {
    const { result } = renderHook(() => useForm());
    const { container } = render(
      <Input label="Test" register={result.current.register} />
    );

    expect(screen.getByRole("textbox")).toHaveAttribute("placeholder", "test");

    expect(container).toMatchSnapshot();
  });

  it("renders an <input> with name", () => {
    const { result } = renderHook(() => useForm());
    const { container } = render(
      <Input name="test" register={result.current.register} />
    );

    expect(screen.getByRole("textbox")).toHaveAttribute("placeholder", "test");

    expect(container).toMatchSnapshot();
  });

  it("render an <input> without register", () => {
    const { container } = render(<Input placeholder="test-input" />);

    expect(screen.getByRole("textbox")).toHaveAttribute(
      "placeholder",
      "test-input"
    );

    expect(container).toMatchSnapshot();
  });
});
