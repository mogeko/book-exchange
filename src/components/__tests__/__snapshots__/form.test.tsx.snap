// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input render an <input> without register 1`] = `
<div>
  <input
    class="input input-bordered"
    placeholder="test-input"
  />
</div>
`;

exports[`Input renders an <input> with label 1`] = `
<div>
  <div
    class="form-control"
  >
    <label
      class="label"
    >
      <span
        class="label-text"
      >
        Test
      </span>
    </label>
    <input
      class="input input-bordered"
      name="test"
      placeholder="test"
    />
  </div>
</div>
`;

exports[`Input renders an <input> with name 1`] = `
<div>
  <div
    class="form-control"
  >
    <input
      class="input input-bordered"
      name="test"
      placeholder="test"
    />
  </div>
</div>
`;
